### PR TITLE
Update code

### LINKED ISSUE #336: Unable to find template "SimpleThingsEntityAuditBundle:Audit:index.html.twig"
| Q                | A
| ---------------- | -----
| Bug report?      | yes
| Feature request? | no
| BC Break report? | yes
| Used Version     | e.g. 1.0.1

I installed the bundle on both Symfony 3 and Symfony 4 versions. Installed successfully, logs are inserting successfully. But when I call the URL - 

http://localhost:3000/audit

it gives the mentioned below error - 

Unable to find template "SimpleThingsEntityAuditBundle:Audit:index.html.twig" (looked into: /home/mycomputer/htdocs/ListManagement-master/app/Resources/views, /home/mycomputer/htdocs/ListManagement-master/vendor/symfony/symfony/src/Symfony/Bridge/Twig/Resources/views/Form).

Please suggest how can we resolve it. I also tried by adding the twig in the mentioned path but still same error.

I need to run it on Symfony 4

Thanks.

### PR DESCRIPTION
**Pull request by wbloszyk**
## Subject

I am targeting this branch, because this change respect BC.

Closes #336

## To do
- [x] Update references to templates
- [x] Update routing sturcture
- [x] Make test finals
- [x] Add type hints to tests
- [x] Using FQCN instead class-string

### CODE REVIEW

File: src/Resources/config/routing.yml [hunk 1]
```diff
@@ -1,33 +1,30 @@
 simple_things_entity_audit_home:
     path: /{page}
+    controller: SimpleThings\EntityAudit\Controller\AuditController::index
```
Comment (wbloszyk) [inline src/Resources/config/routing.yml:3]:
```suggestion
    controller: SimpleThings\EntityAudit\Controller\AuditController::indexAction
```
Shouldnt I add full method name? @sonata-project/contributors

File: src/Controller/AuditController.php [hunk 2]
```diff
@@ -36,7 +36,7 @@ public function indexAction($page = 1)
         $reader = $this->getAuditReader();
         $revisions = $reader->findRevisionHistory(20, 20 * ($page - 1));
 
-        return $this->render('SimpleThingsEntityAuditBundle:Audit:index.html.twig', [
+        return $this->render('@SimpleThingsEntityAudit\Audit\index.html.twig', [
```
Comment (phansys) [inline src/Controller/AuditController.php:39]: Why are we using backslashes?

**Review by phansys — COMMENTED** [review]
Could we use this PR for other pedantic changes? Like making final the tests classes, or using the `::class` pseudo constant where there are hardcoded references to some class names.

File: src/Controller/AuditController.php (hunk 2 above)
Comment (wbloszyk) [reply src/Controller/AuditController.php]: 🤦🏻 I do it again. I dont know, mybe I love backslashes. Fixed.

**Comment by wbloszyk** [comment]
> Could we use this PR for other pedantic changes? Like making final the tests classes, or using the `::class` pseudo constant where there are hardcoded references to some class names.

Shure.

File: tests/Fixtures/Core/AnimalAudit.php [hunk 3]
```diff
@@ -12,10 +12,15 @@
  */
 abstract class AnimalAudit
 {
-    /** @ORM\Id @ORM\Column(type="integer") @ORM\GeneratedValue */
+    /**
+     * @ORM\Id @ORM\Column(type="integer")
```
Comment (VincentLanglet) [inline tests/Fixtures/Core/AnimalAudit.php:16]:
```suggestion
     * @ORM\Id
     * @ORM\Column(type="integer")
```

File: tests/Fixtures/Core/ArticleAudit.php [hunk 4]
```diff
@@ -7,21 +7,32 @@
 /**
  * @ORM\Entity
  */
-class ArticleAudit
+final class ArticleAudit
 {
-    /** @ORM\Id @ORM\Column(type="integer") @ORM\GeneratedValue */
+    /**
+     * @ORM\Id @ORM\Column(type="integer")
```
Comment (VincentLanglet) [inline tests/Fixtures/Core/ArticleAudit.php:13]:
```suggestion
     * @ORM\Id
     * @ORM\Column(type="integer")
```

File: tests/Fixtures/Core/PetAudit.php [hunk 5]
```diff
@@ -12,10 +12,15 @@
  */
 abstract class PetAudit
 {
-    /** @ORM\Id @ORM\Column(type="integer") @ORM\GeneratedValue */
+    /**
+     * @ORM\Id @ORM\Column(type="integer")
```
Comment (VincentLanglet) [inline tests/Fixtures/Core/PetAudit.php:16]:
```suggestion
     * @ORM\Id
     * @ORM\Column(type="integer")
```

File: tests/Fixtures/Core/ProfileAudit.php [hunk 6]
```diff
@@ -7,15 +7,23 @@
 /**
  * @ORM\Entity
  */
-class ProfileAudit
+final class ProfileAudit
 {
-    /** @ORM\Id @ORM\Column(type="integer") @ORM\GeneratedValue */
+    /**
+     * @ORM\Id @ORM\Column(type="integer")
```
Comment (VincentLanglet) [inline tests/Fixtures/Core/ProfileAudit.php:13]:
```suggestion
     * @ORM\Id
     * @ORM\Column(type="integer")
```

File: tests/Fixtures/Core/UserAudit.php [hunk 7]
```diff
@@ -7,14 +7,21 @@
 /**
  * @ORM\Entity
  */
-class UserAudit
+final class UserAudit
 {
-    /** @ORM\Id @ORM\Column(type="integer") @ORM\GeneratedValue */
+    /**
+     * @ORM\Id @ORM\Column(type="integer")
```
Comment (VincentLanglet) [inline tests/Fixtures/Core/UserAudit.php:13]:
```suggestion
     * @ORM\Id
     * @ORM\Column(type="integer")
```

File: tests/Fixtures/Issue/ConvertToPHPEntity.php [hunk 8]
```diff
@@ -7,15 +7,20 @@
 /**
  * @ORM\Entity()
  */
-class ConvertToPHPEntity
+final class ConvertToPHPEntity
 {
-    /** @ORM\Id @ORM\Column(type="integer") @ORM\GeneratedValue(strategy="AUTO") */
-    protected $id;
+    /**
+     * @ORM\Id @ORM\Column(type="integer")
```
Comment (VincentLanglet) [inline tests/Fixtures/Issue/ConvertToPHPEntity.php:13]:
```suggestion
     * @ORM\Id
     * @ORM\Column(type="integer")
```

File: tests/Fixtures/Issue/Issue111Entity.php [hunk 9]
```diff
@@ -9,18 +9,25 @@
...
-
-    /** @ORM\Column */
-    protected $status;
-
-    /** @ORM\Column(type="datetime", nullable=true, name="deleted_at") */
-    protected $deletedAt;
-
-    public function getId()
+    /**
+     * @ORM\Id @ORM\Column(type="integer")
```
Comment (VincentLanglet) [inline tests/Fixtures/Issue/Issue111Entity.php:15]:
```suggestion
     * @ORM\Id
     * @ORM\Column(type="integer")
```

File: tests/Fixtures/Issue/Issue198Car.php [hunk 10]
```diff
@@ -1,37 +1,41 @@
...
  * @ORM\Entity()
  */
-class Issue198Car
+final class Issue198Car
 {
-    /** @ORM\Id @ORM\Column(type="integer") @ORM\GeneratedValue(strategy="AUTO") */
-    protected $id;
-    
+    /**
+     * @ORM\Id @ORM\Column(type="integer")
```
Comment (VincentLanglet) [inline tests/Fixtures/Issue/Issue198Car.php:13]:
```suggestion
     * @ORM\Id
     * @ORM\Column(type="integer")
```

File: tests/Fixtures/Issue/Issue198Owner.php [hunk 11]
```diff
@@ -7,44 +8,47 @@
...
  * @ORM\Entity()
  */
-class Issue198Owner
+final class Issue198Owner
 {
-    /** @ORM\Id @ORM\Column(type="integer") @ORM\GeneratedValue(strategy="AUTO") */
-    protected $id;
-    
+    /**
+     * @ORM\Id @ORM\Column(type="integer")
```
Comment (VincentLanglet) [inline tests/Fixtures/Issue/Issue198Owner.php:14]:
```suggestion
     * @ORM\Id
     * @ORM\Column(type="integer")
```

File: tests/Fixtures/Issue/Issue87AbstractProject.php [hunk 12]
```diff
@@ -13,52 +13,59 @@
  */
 abstract class Issue87AbstractProject
 {
-    /** @ORM\Id @ORM\Column(type="integer") @ORM\GeneratedValue(strategy="AUTO") */
-    protected $id;
+    /**
+     * @ORM\Id @ORM\Column(type="integer")
```
Comment (VincentLanglet) [inline tests/Fixtures/Issue/Issue87AbstractProject.php:17]:
```suggestion
     * @ORM\Id
     * @ORM\Column(type="integer")
```

File: tests/Fixtures/Issue/Issue87Organization.php [hunk 13]
```diff
@@ -4,13 +4,18 @@
...
-class Issue87Organization
+/**
+ * @ORM\Entity
+ */
+final class Issue87Organization
 {
-    /** @ORM\Id @ORM\Column(type="integer") @ORM\GeneratedValue(strategy="AUTO") */
-    protected $id;
+    /**
+     * @ORM\Id @ORM\Column(type="integer")
```
Comment (VincentLanglet) [inline tests/Fixtures/Issue/Issue87Organization.php:13]:
```suggestion
     * @ORM\Id
     * @ORM\Column(type="integer")
```

File: tests/Fixtures/Issue/Issue198Owner.php [hunk 14]
```diff
@@ -7,44 +8,47 @@
...
             $this->cars[] = $car;
         }
     }
-    
+
     public function removeCar(Issue198Car $car)
     {
         $this->cars->removeElement($car);
     }
-
```
Comment (VincentLanglet) [inline tests/Fixtures/Issue/Issue198Owner.php:47]: Why this was not fix by the linter ?

The lint action was green, and the rules
```
no_trailing_whitespace
no_whitespace_in_blank_line
```
are included in the Symfony ruleset.

Comment (VincentLanglet) [reply tests/Fixtures/Issue/Issue198Owner.php:47]: Oh, it's in the Fixture folder

Comment (VincentLanglet) [reply tests/Fixtures/Issue/Issue198Owner.php:47]: I made https://github.com/sonata-project/dev-kit/pull/1418

**Comment by wbloszyk** [comment]
I got some confusing error. 
```
Proxies\__CG__\SimpleThings\EntityAudit\Tests\Fixtures\Issue\DuplicateRevisionFailureTestSecondaryOwner
::setPrimaryOwner(DuplicateRevisionFailureTestPrimaryOwner $owner) 

must be compatible with 

SimpleThings\EntityAudit\Tests\Fixtures\Issue\DuplicateRevisionFailureTestSecondaryOwner
::setPrimaryOwner(DuplicateRevisionFailureTestPrimaryOwner $owner): void 
```

IMO It is doctrine and we should bump it. Do you know what should be bump? @sonata-project/contributors

**Review by VincentLanglet — APPROVED** [review]

File: src/Resources/config/routing.yml [hunk 15]
```diff
@@ -1,33 +1,30 @@
 simple_things_entity_audit_home:
     path: /{page}
+    controller: SimpleThings\EntityAudit\Controller\AuditController::indexAction
```
Comment (jordisala1991) [inline src/Resources/config/routing.yml:3]: Are you sure the controllers are defined like services here?

Comment (wbloszyk) [reply src/Resources/config/routing.yml:3]: It is instance of `ContainerAwareInterface` so it should be trait like FQCN and autogenerate like with `SimpleThingsEntityAuditBundle:Audit:index`. I will check it.

Comment (wbloszyk) [reply src/Resources/config/routing.yml:3]: @jordisala1991 
After check it is working.

**Review by jordisala1991 — APPROVED** [review]

**Comment by wbloszyk** [comment]
@phansys Can we merge it?

File: src/Exception/NotAuditedException.php [hunk 16]
```diff
@@ -19,7 +19,7 @@ public function __construct($className)
     {
         parent::__construct($className, null, null);
         $this->message = sprintf(
-            "Class '$className' is not audited.",
+            "Class '%s' is not audited.",
```
Comment (phansys) [inline src/Exception/NotAuditedException.php:22]:
```suggestion
            'Class "%s" is not audited.',
```

File: tests/Fixtures/Core/UserAudit.php [hunk 17]
```diff
@@ -9,43 +20,48 @@
...
     private $name;
 
-    /** @ORM\OneToOne(targetEntity="ProfileAudit", mappedBy="user") */
+    /**
+     * @ORM\OneToOne(targetEntity="ProfileAudit", mappedBy="user")
+     */
     private $profile;
 
-    function __construct($name)
+    public function __construct($name)
```
Comment (phansys) [inline tests/Fixtures/Core/UserAudit.php:39]:
```suggestion
    public function __construct(string $name)
```

File: tests/Utils/SimpleDiffTest.php [hunk 18]
```diff
@@ -16,7 +16,7 @@
 use PHPUnit\Framework\TestCase;
 use SimpleThings\EntityAudit\Utils\SimpleDiff;
 
-class SimpleDiffTest extends TestCase
+final class SimpleDiffTest extends TestCase
```
Comment (phansys) [inline tests/Utils/SimpleDiffTest.php:19]: Could you please add the return type declaration to `dataDiff()` method and the argument type declarations to `testDiff()`?

File: tests/CoreTest.php [hunk 19]
```diff
@@ -137,11 +143,12 @@ public function testFindNoRevisionFound(): void
         $reader = $this->auditManager->createAuditReader($this->em);
 
         $this->expectException(NoRevisionFoundException::class);
-        $this->expectExceptionMessage(
-            "No revision of class 'SimpleThings\\EntityAudit\\Tests\\Fixtures\\Core\\UserAudit' (1) was found at revision 1 or before. The entity did not exist at the specified revision yet."
-        );
+        $this->expectExceptionMessage(sprintf(
+            "No revision of class '%s' (1) was found at revision 1 or before. The entity did not exist at the specified revision yet.",
```
Comment (phansys) [inline tests/CoreTest.php:147]: Could you please propagate this change for all the exception messages in order to use single quotes where possible?

```suggestion
            'No revision of class "%s" (1) was found at revision 1 or before. The entity did not exist at the specified revision yet.',
```

**Review by phansys — CHANGES_REQUESTED** [review]

File: tests/Utils/SimpleDiffTest.php (hunk 18 above)
Comment (wbloszyk) [reply tests/Utils/SimpleDiffTest.php:19]: Done.

**Review by phansys — APPROVED** [review]

**Review by VincentLanglet — APPROVED** [review]

**Pull request merged**